pilot/pkg/xds/extension_discovery.go

```diff
@@ -0,0 +1,131 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package xds
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	core "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	discovery "github.com/envoyproxy/go-control-plane/envoy/service/discovery/v3"
+	extensionservice "github.com/envoyproxy/go-control-plane/envoy/service/extension/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/cache/v3"
+	"github.com/envoyproxy/go-control-plane/pkg/server/v3"
+	"google.golang.org/grpc/codes"
+	status "google.golang.org/grpc/status"
+	"google.golang.org/protobuf/proto"
+
+	networking "istio.io/api/networking/v1alpha3"
+	"istio.io/istio/pilot/pkg/model"
+)
+
+const (
+	// ExtensionConfigTypeURL for extension configs.
+	ExtensionConfigTypeURL = "type.googleapis.com/envoy.config.core.v3.TypedExtensionConfig"
+)
+
+// ExtensionServer is the main server instance.
+type ExtensionServer struct {
+	sync.Mutex
+	server.Server
+	server.CallbackFuncs
+	cache   *cache.LinearCache
+	configs map[string]*core.TypedExtensionConfig
+}
+
+var _ extensionservice.ExtensionConfigDiscoveryServiceServer = &ExtensionServer{}
+
+func NewExtensionServer() *ExtensionServer {
+	out := &ExtensionServer{}
+	out.cache = cache.NewLinearCache(ExtensionConfigTypeURL, cache.WithVersionPrefix(time.Now().Format(time.RFC3339)+"/"))
```

Comment:
I am not sure the benefit of running two different XDS implementations (go-control-plane and pilot's XDS handling). This seems like it will lead to a lot of complexity/odd behavioral differences without much benefit

Reply:
I'm not familiar with pilot code. If you can give me a stub for ECDS, I can put it there. But this accomplishes the same in <100 lines of code.

Reply:
You just need to implement https://github.com/istio/istio/blob/13b49016b6f1fcd6f95b5a0048f4792affd6b44b/pilot/pkg/model/context.go#L145. I would refer to sds.go, which is newly added and required pretty minimal code changes. The major difference is you want to use ECDS split, not ADS. I think this can be fairly simply by splitting the code in https://github.com/istio/istio/blob/2a8234f6de8cefc16cedee6de567c34048c2f8ad/pilot/pkg/xds/ads.go#L185 into a different function. Both ADS and ECDS implement `DiscoveryStream` so we can pretty much have a single call out to some common function, possible with a minor tweak to handle the fact that TypeUrl is optional in non ADS I think?

Reply:
Sorry, I didn't mean just the stub. I'm concerned about the payload size in ECDS. Each resource can be 10MB with pre-compiled Wasm code. Which means we'd need to serve up to 100*10 = 1GB of data. So we should unnecessary copies (per-connection, per config update, etc), but that requires understanding the whole data flow, which is hard to decipher.

Reply:
I am not sure I completely understand the issue with payload size. You can send the same exact response as with gcp. For example, marshaled resources can easily be cached: https://github.com/istio/istio/blob/fb05b8b5606d489387e76b45ec36ba36703f546f/pilot/pkg/xds/sds.go#L137-L141 and you choose what resources to respond so it has no impact there.

Reply:
where is the code to validate the pod's service account token? or use mTLS if using provisioned certs..

Reply:
That needs to be refactored into a gRPC middleware. They way it is written is not re-usable.